HACKING.md

```diff
@@ -0,0 +1,409 @@
+
+### On Linux
+
+Install pyenv: 
+
+    git clone https://github.com/pyenv/pyenv.git ~/.pyenv
+    export PATH="$HOME/.pyenv/bin:$PATH"
+    git clone https://github.com/pyenv/pyenv-virtualenv.git $(pyenv root)/plugins/pyenv-virtualenv
+
+### On Mac
+
+Install pyenv using brew:
+
+    brew update
+    brew install pyenv
+    brew install pyenv-virtualenv
+
+---
+
+Now add pyenv to your `.bashrc`:
+
+    export PATH="$HOME/.pyenv/bin:$PATH" >> ~/.bashrc
+    pyenv init 2>> ~/.bashrc
+    pyenv virtualenv-init 2>> ~/.bashrc
+
+For other shells, see [the pyenv instructions](https://github.com/pyenv/pyenv#basic-github-checkout).
+
+Go ahead and restart your shell. Now you should be able to install a new
+version of Python:
+
+    pyenv install 3.7.0
+
+If that fails, don't panic! You're probably just missing some build
+dependencies. Check out [pyenv common build
+problems](https://github.com/pyenv/pyenv/wiki/Common-build-problems).
+
+Now that you've got your version of Python installed, you can create a new
+virtual environment with this command:
+
+    pyenv virtualenv 3.7.0 salt
+
+Then activate it:
+
+    pyenv activate salt
+
+Sweet! Now you're ready to clone salt so you can start hacking away!
+
+### Cloning Salt
+
+Clones are so shallow. Well, this one is anyway:
+
+    git clone --depth=1 --origin salt https://github.com/saltstack/salt.git
+
+This creates a shallow clone of salt, which should be fast. Most of the time
+that's all you'll need, and you can start building out other commits as you go.
+If you *really* want all 108,300+ commits you can just run `git fetch
+--unshallow`. Then go make a sandwich because it's gonna be a while.
+
+### Fork Salt
+
+You're also going to want to head over to GitHub and create your own [fork of
+salt](https://github.com/saltstack/salt/fork). Once you've got that setup you
+can add it as a remote:
+
+    git remote add yourname <YOUR SALT REMOTE>
+
+If you use your name to refer to your fork, and `salt` to refer to the official
+Salt repo you'll never get `upstream` or `origin` confused.
+
+### `pre-commit` and `nox` Setup
+
+Here at Salt we use [pre-commit](https://pre-commit.com/) and
+[nox](https://nox.thea.codes/en/stable/) to make it easier for contributors to
+get quick feedback. Nox enables us to run multiple different test
+configurations, as well as other common tasks. You can think of it as Make with
+superpowers. Pre-commit does what it sounds like - it configures some Git
+pre-commit hooks to run `black` for formatting, `isort` for keeping our imports
+sorted, and `pylint` to catch issues like unused imports, among others. You can
+easily install them in your virtualenv with:
+
+    python -m pip install pre-commit nox
+    pre-commit install
+
+Now before each commit, it will ensure that your code at least *looks* right
+before you open a pull request. And with that step, it's time to start hacking
+on Salt!
+
+
+## Selecting an Issue
+
```

Comment:
**Question:** Should we also have a section on opening issues?

Reply:
Actually... yeah. That's an excellent point - since we're making the statement that /all/ contributions are contributions, I should definitely write something up about what it means to open a /good/ issue :+1: